File: java/Ice/bidir/CallbackSenderI.java
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

import com.zeroc.demos.Ice.bidir.Demo.*;
import java.util.concurrent.*;

class CallbackSenderI implements CallbackSender
{
    @Override
    synchronized public void addClient(CallbackReceiverPrx client, com.zeroc.Ice.Current current)
    {
        System.out.println("adding client `" + com.zeroc.Ice.Util.identityToString(client.ice_getIdentity()) + "'");
        _clients.add(client.ice_fixed(current.con));
    }

    public void destroy()
    {
        System.out.println("destroying callback sender");
        _executorService.shutdown();
        try
        {
            _executorService.awaitTermination(10, TimeUnit.SECONDS);
        }
        catch(InterruptedException interrupted)
        {
            // ignored
        }
    }

    public void start()
    {
        _executorService.scheduleAtFixedRate(() -> invokeCallback(), 2, 2, TimeUnit.SECONDS);
    }

    synchronized private void invokeCallback()
    {
        if(!_clients.isEmpty())
        {
            _num++;

            //
            // Invoke callback on all clients; it's safe to do it within the synchronization
            // because Ice guarantees these async invocations never block the calling thread.
            //
            // We use whenCompleteAsync instead of whenComplete to ensure that if the
            // call completes immediately we delay the removeClient until after the iteration.
            //
            for(CallbackReceiverPrx p : _clients)
            {
                try
                {
                    p.callbackAsync(_num).whenCompleteAsync(
                        (r, t) ->
                        {
                            if(t != null)
                            {
                                removeClient(p, t);
                            }
                        },
                        _executorService);
                }
                catch(com.zeroc.Ice.CommunicatorDestroyedException e)
                {
                    //
                    // We're shutting down
                    //
                    break;
                }
            }
        }
    }

    synchronized private void removeClient(CallbackReceiverPrx p, Throwable t)
    {
        if(_clients.remove(p))
        {
            System.err.println("removing client `" + com.zeroc.Ice.Util.identityToString(p.ice_getIdentity()) + "':");
            t.printStackTrace();
        }
    }

    private ScheduledExecutorService _executorService = Executors.newSingleThreadScheduledExecutor();
    private java.util.List<CallbackReceiverPrx> _clients = new java.util.ArrayList<>();
    private int _num = 0;
}

File: csharp/Ice/bidir/CallbackSenderI.cs
// Copyright (c) ZeroC, Inc.

using Demo;

internal class CallbackSenderI(Ice.Communicator communicator) : CallbackSenderDisp_
{
    public void Destroy()
    {
        lock (this)
        {
            Console.Out.WriteLine("destroying callback sender");
            _destroy = true;
            System.Threading.Monitor.Pulse(this);
        }
    }

    public override void addClient(CallbackReceiverPrx client, Ice.Current current)
    {
        lock (this)
        {
            Console.Out.WriteLine("adding client `" + Ice.Util.identityToString(client.ice_getIdentity()) + "'");
            _clients.Add((CallbackReceiverPrx)client.ice_fixed(current.con));
        }
    }

    public void Run()
    {
        int num = 0;
        lock (this)
        {
            while (!_destroy)
            {
                System.Threading.Monitor.Wait(this, 2000);

                if (!_destroy && _clients.Count > 0)
                {
                    ++num;

                    // Invoke callback on all clients; it's safe to do it with _mutex locked
                    // because Ice guarantees these async invocations never block the calling thread.
                    //
                    // The exception callback, if called, is called by an Ice thread, and never the
                    // calling thread.
                    foreach (var c in _clients)
                    {
                        c.callbackAsync(num).ContinueWith(t =>
                        {
                            if (t.Exception != null)
                            {
                                RemoveClient(c, t.Exception.InnerException);
                            }
                        });
                    }
                }
            }
        }
    }

    private void RemoveClient(CallbackReceiverPrx c, Exception ex)
    {
        lock (this)
        {
            if (_clients.Remove(c))
            {
                Console.Error.WriteLine("removing client `" + Ice.Util.identityToString(c.ice_getIdentity()) + "':\n" +
                                        ex);
            }
        }
    }

    private Ice.Communicator _communicator = communicator;
    private bool _destroy = false;
    private List<CallbackReceiverPrx> _clients = new List<CallbackReceiverPrx>();
}
